src/libraries/Microsoft.PowerFx.Core/Utils/TextFormatUtils.cs

```diff
@@ -0,0 +1,79 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Text.RegularExpressions;
+using Microsoft.PowerFx.Core.Errors;
+using Microsoft.PowerFx.Core.Localization;
+using System.Xml.Linq;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx.Core.Utils
+{
+    internal sealed class TextFormatUtils
+    {
+        static public DateTime BaseDateTime => new DateTime(1899, 12, 30, 0, 0, 0, 0);
+
+        public static bool IsValidFormatArg(string formatArg, out bool hasDateTimeFmt, out bool hasNumericFmt)
+        {
+            // Verify statically that the format string doesn't contain BOTH numeric and date/time
+            // format specifiers. If it does, that's an error according to Excel and our spec.
+
+            // But firstly skip any locale-prefix
+            if (formatArg.StartsWith("[$-"))
+            {
+                var end = formatArg.IndexOf(']', 3);
+                if (end > 0)
+                {
+                    formatArg = formatArg.Substring(end + 1);
+                }
+            }
+
+            hasDateTimeFmt = formatArg.IndexOfAny(new char[] { 'm', 'M', 'd', 'D', 'y', 'Y', 'h', 'H', 's', 'S', 'a', 'A', 'p', 'P' }) >= 0;
+            hasNumericFmt = formatArg.IndexOfAny(new char[] { '0', '#' }) >= 0;
+            if (hasDateTimeFmt && hasNumericFmt)
+            {
+                // Check if the date time format contains '0's after the seconds specifier, which
+                // is used for fractional seconds - in which case it is valid
+                var formatWithoutZeroSubseconds = Regex.Replace(formatArg, @"[sS]\.?(0+)",
```

Comment:
>@"[sS]\.?(0+)" [](http://example.com/codeflow?start=75&length=14)

Make this a private readonly static member, so that this regex is only compiled once.

Reply:
This regex is responsible for getting rid of the ambiguity between datatime format and numeric format. I'm under the impression that this regex is checking if we have seconds char (sS) followed by dot + zero(s). A "0" (zero) under this circumstances means this is a datetime format.

Reply:
Yes, you're right. But it can be done more efficiently without having specific logic to replace each captured group. Also, we should avoid using the static Regex.Replace(text, regularExpression), instead defining the regexes once as static members so that they are not recomputed every time the function is run.